libass/ass_coretext.c

```diff
@@ -270,7 +274,18 @@ static char *get_fallback(void *priv, ASS_Library *lib,
         return NULL;
     }
 
-    CTFontRef fb = CTFontCreateForString(font, r, CFRangeMake(0, 1));
+    CTFontRef fb = NULL;
+#if __has_builtin(__builtin_available) && defined(MAC_OS_X_VERSION_10_9)
+    if (__builtin_available(macOS 10.9, iOS 7.0, *)) {
```

Comment:
It turns out this symbol isn’t actually declared in the C headers until the 10.15 SDK. As discussed on IRC, let’s declare it ourselves and hope for the best.

Further, as discussed in https://github.com/libass/libass/pull/664#discussion_r1119732310, in https://github.com/libass/libass/pull/664/commits/77e4bfb977ebd8bcde56f21da88760046a025f39 and on IRC, the following will restore compatibility with GCC and older Clang, which are viable compilers on 10.9, and make the `__builtin_available` polyfill redundant:

```suggestion
#if __MAC_OS_X_VERSION_MAX_ALLOWED >= 1090 || __IPHONE_OS_VERSION_MAX_ALLOWED >= 70000
    CTFontRef CTFontCreateForStringWithLanguage(CTFontRef currentFont, CFStringRef string, CFRange range, CFStringRef language) __OSX_AVAILABLE_STARTING(__MAC_10_9, __IPHONE_7_0);
    if ((intptr_t) &CTFontCreateForStringWithLanguage) {
```

plus `#include <Availability.h>` (or we can rely on the fact that it already happens to be included transitively, but I’d slightly prefer an explicit include),

assuming I can trust the claim in Availability.h that `__MAC_OS_X_VERSION_MAX_ALLOWED` is undefined when targeting iOS.

Reply:
This introduces a `-Wpartial-availability` warning when targeting pre-10.9, which is often configured as an error.

Reply:
Current docs say `-Wpartial-availability` is an alias for `-Wunguarded-availability` which is the exact flag that warns about `availability`, and as you and I have both mentioned previously, it is only enabled by default in its `-Wunguarded-availability-new` which affects 10.13+ APIs, not 10.9 APIs. Our Makefile doesn’t enable it, either. This is no different from #664.

Unless perhaps you actually mean some other warning that pertains to the declaration rather than the invocation?

If you want us to support build setups that explicitly add `-Werror=unguarded-availability` to extra cflags, we need completely separate compilation paths (or a macro) for new Clang and all other compilers.

Reply:
Re-pushed with a new `CHECK_AVAILABLE` macro, which should work fine for both new and old compilers.